### PR TITLE
⚡️ Add caching layer for rendered templates

### PR DESCRIPTION
**Pull request by matthiasrohmer**
Currently doing another another round of testing on staging. Open questions:

1) Which strategy do we want to use to pick instances? Do we even configure multiple per region?
https://github.com/ampproject/amp.dev/blob/4f22ac660423b41e5995cd51f1223be6e89c6cd3/platform/lib/utils/pageCache.js#L101-L103

2) Should the pageCache fallback to `lru-cache` if no instance can be found?

3) What's a expiration time that makes sense? I've gone with a day for now as we are doing cron-deploys every night:
https://github.com/ampproject/amp.dev/blob/4f22ac660423b41e5995cd51f1223be6e89c6cd3/platform/lib/utils/pageCache.js#L35

### CODE REVIEW

File: platform/lib/routers/growPages.js [hunk 1]
```diff
@@ -291,6 +299,9 @@ growPages.get(/^(.*\/)?([^\/\.]+|.+\.html|.*\/|$)$/, async (req, res, next) => {
     signale.error('[OPTIMIZER]', e);
   }
 
+  // Cache the optimized and rendered page
+  pageCache.set(req.originalUrl, renderedTemplate);
```
Comment (sebastianbenz) [inline platform/lib/routers/growPages.js]: Do this after sending the page content

Comment (sebastianbenz) [reply platform/lib/routers/growPages.js]: Don't we need to encode the current version into the cache key?

Comment (matthiasrohmer) [reply platform/lib/routers/growPages.js]: Yep that's done in `pageCache` itself.

**Comment by matthiasrohmer** [comment]
😎 

```
2019-12-02 16:09:04.000 MEZ
[CRITICAL_TIMING] amp-dev-staging.appspot.com/ 1.584s [200]
[TIMING] amp-dev-staging.appspot.com/ 0.014s [200]
```

Some more testing I did with artillery:

### Without cache
Bildschirmfoto 2019-12-03 um 14 26 10

```
All virtual users finished
Summary report @ 14:14:24(+0100) 2019-12-03
  Scenarios launched:  1560
  Scenarios completed: 1560
  Requests completed:  1647
  RPS sent: 2.11
  Request latency:
    min: 183.3
    max: 70681.5
    median: 487.7
    p95: 853.1
    p99: 1445
  Scenario counts:
    Surf through various pages: 1560 (100%)
  Codes:
    200: 1560
    301: 87
```

### With cache
Bildschirmfoto 2019-12-03 um 13 37 29

```
All virtual users finished
Elapsed time: 13 minutes, 1 second
Summary report @ 13:33:59(+0100) 2019-12-03
  Scenarios launched:  1560
  Scenarios completed: 1558
  Requests completed:  1647
  RPS sent: 2.11
  Request latency:
    min: 124.8
    max: 1433.5
    median: 290.6
    p95: 594.5
    p99: 759.3
  Scenario counts:
    Surf through various pages: 1560 (100%)
  Codes:
    200: 1560
    301: 89
  Errors:
    ECONNRESET: 2
```

File: platform/lib/utils/pageCache.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    lru = new LRU({
+      max: LRU_MAX_ITEMS,
+    });
+  } else {
+    const instance = instances[0];
+    console.log('[PAGE_CACHE]: About to connect to Redis at', instance.port, instance.host);
+    try {
+      redis = new Redis(instance.port, instance.host);
+    } catch(e) {
+      console.error('[PAGE_CACHE]: Connecting to Redis failed', e);
```
Comment (sebastianbenz) [inline platform/lib/utils/pageCache.js]: we should  fallback to lru as well

**Pull request merged**